### ISSUE TITLE
C++ Build Failure with MSVC on cxx/1.0.11+

### ISSUE THREAD

**Issue by david-cattermole** [dtolnay/cxx#568]
Hello dtolnay,

I am having problems building my `cxx-demo-example` repo on Windows (with MSVC 19.00.24215.1 and Microsoft Visual Studio 14.0) with `cxx/1.0.11` up to the latest release (currently `cxx/1.0.16`). `cxx/1.0.10` works perfectly and is the last version I can successfully use. I expect the problem is introduced with #538 and the Vec random access iterator.

I have not had this problem on Linux (with GCC 6.3.1), and you can see the build output here:
https://github.com/david-cattermole/cxx-demo-example/blob/59c9a6f7eefca90cad7d7f60330673b785d2e36c/output_gcc_6_3_1.log

I am attempting to diagnose and debug the problem, but I felt you should know about the problem.

To test this, you can clone this:
https://github.com/david-cattermole/cxx-demo-example

...and then run this:
```
> cd C:\path\to\project\root\
> build_windows64.bat
Package Root: C:\Users\catte\dev\cxx-demo-example\
thread 'main' panicked at 'Unable to find mmscenegraph for "C:\\Users\\catte\\dev\\cxx-demo-example\\Cargo.toml"', C:\Users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cbindgen-0.14.2\src\bindgen\cargo\cargo.rs:88:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
   Compiling cc v1.0.62
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.54
   Compiling cxxbridge-flags v1.0.11
   Compiling libc v0.2.80
   Compiling link-cplusplus v1.0.4
   Compiling cxx v1.0.11
   Compiling quote v1.0.7
error: failed to run custom build command for `cxx v1.0.11`

Caused by:
  process didn't exit successfully: `C:\Users\catte\dev\cxx-demo-example\target\release\build\cxx-6a0581fa8fa2d641\build-script-build` (exit code: 1)
  --- stdout
  TARGET = Some("x86_64-pc-windows-msvc")
  OPT_LEVEL = Some("3")
  HOST = Some("x86_64-pc-windows-msvc")
  CXX_x86_64-pc-windows-msvc = None
  CXX_x86_64_pc_windows_msvc = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-pc-windows-msvc = None
  CXXFLAGS_x86_64_pc_windows_msvc = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  DEBUG = Some("false")
  CXX_x86_64-pc-windows-msvc = None
  CXX_x86_64_pc_windows_msvc = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-pc-windows-msvc = None
  CXXFLAGS_x86_64_pc_windows_msvc = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  running: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\x86_amd64\\cl.exe" "-nologo" "-MD" "-O2" "-Brepro" "-W4" "-FoC:\\Users\\catte\\dev\\cxx-demo-example\\target\\release\\build\\cxx-9adf7102b1f9cf97\\out\\src/cxx.o" "-c" "src/cxx.cc"
  cxx.cc
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\xlocale(341): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(768): warning C4346: 'rust::cxxbridge1::Vec<T>::iterator::difference_type': dependent name is not a type
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(768): note: prefix with 'typename' to indicate a type
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(768): error C2146: syntax error: missing ')' before identifier 'n'
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(774): error C2143: syntax error: missing ';' before '{'
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(774): error C2447: '{': missing function header (old-style formal list?)
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(801): warning C4346: 'rust::cxxbridge1::Vec<T>::iterator::difference_type': dependent name is not a type
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(801): note: prefix with 'typename' to indicate a type
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(801): error C2146: syntax error: missing ')' before identifier 'n'
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(808): error C2143: syntax error: missing ';' before '{'
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(808): error C2447: '{': missing function header (old-style formal list?)
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(815): warning C4346: 'rust::cxxbridge1::Vec<T>::iterator::difference_type': dependent name is not a type
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(815): note: prefix with 'typename' to indicate a type
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(815): error C2146: syntax error: missing ')' before identifier 'n'
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(823): error C2143: syntax error: missing ';' before '{'
  c:\users\catte\.cargo\registry\src\github.com-1ecc6299db9ec823\cxx-1.0.11\src\../include/cxx.h(823): error C2447: '{': missing function header (old-style formal list?)
  src/cxx.cc(293): warning C4996: 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
  C:\Program Files (x86)\Windows Kits\10\include\10.0.10586.0\ucrt\string.h(335): note: see declaration of 'strncpy'
  exit code: 2

  --- stderr


  error occurred: Command "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\x86_amd64\\cl.exe" "-nologo" "-MD" "-O2" "-Brepro" "-W4" "-FoC:\\Users\\catte\\dev\\cxx-demo-example\\target\\release\\build\\cxx-9adf7102b1f9cf97\\out\\src/cxx.o" "-c" "src/cxx.cc" with args "cl.exe" did not execute successfully (status code exit code: 2).


warning: build failed, waiting for other jobs to finish...
error: build failed
```

Thanks,
David

**Comment by dtolnay** [comment]
I put in some additional `typename` in #569. Would it be possible to confirm whether that solves this? I guess it's fixed in newer Visual C++ versions because I was not able to reproduce in a Windows VM or a cursory attempt with GitHub Actions (81d89ea41cff54754d7ad73e883ba8729d46c9d6) but I would accept a PR with a CI build that would have caught this.

**Comment by david-cattermole** [comment]
I can confirm #569 fixes the problem, thank you!

I will look at adding a CI test that can catch this type of problem.

Feel free to close if you'd like. I will add a new PR when/if I can make a test to catch it in the future.

**Comment by dtolnay** [comment]
Published the fix in 1.0.17.

**Comment by david-cattermole** [comment]
Hello @dtolnay, 

I researched this problem (a few weeks ago) but it appears the bug/behaviour was fixed/changed in later versions of MSVC, and there is no way to trigger the old behaviour unless downgrading to an older compiler.

Unfortunately GitHub Actions do not support Visual Studio 2015, only 2017+ so I am not able to reproduce this with GitHub actions.
https://github.com/actions/virtual-environments/blob/main/images/win/Windows2016-Readme.md#visual-studio-enterprise-2017

Just letting you know where I got to with this.

Thanks DavidC.
